lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -451,6 +452,36 @@ resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE, DeclContext *DC) {
       diagnose(Loc, diag::use_unresolved_identifier, Name,
                UDRE->getName().isOperator())
         .highlight(Loc);
+
+      const char *buffer = Name.get();
+      std::string expectedIdentifier = std::string();
+      bool isConfused = false;
+      uint32_t codepoint;
+      int offset = 0;
+      while ((codepoint = validateUTF8CharacterAndAdvance(buffer,
+                                                          buffer +
+                                                          strlen(buffer)))
+             != ~0U) {
+        int length = (buffer - Name.get()) - offset;
+        uint32_t expectedCodepoint;
+        if ((expectedCodepoint =
+             confusable::tryConvertConfusableCharacterToASCII(codepoint))) {
+          isConfused = true;
+          expectedIdentifier += (char)expectedCodepoint;
+        } else {
+          expectedIdentifier += (char)codepoint;
+        }
+
+        offset += length;
+      }
+      
+      if (isConfused) {
```

Comment:
So, to recover I would call `lookupUnqualified` again here and see if you get results back. If you do get results, you can diagnose this as a single `ERROR` with a fix-it and then continue in this function. If you *don't* get any results, I'm not sure we should even say it's confused; using the ASCII characters won't improve anything.

(Well, maybe it will because maybe you've confused characters *and* juxtaposed operators, but that seems kind of unlikely.)